### ISSUE TITLE
R.ifElse - R.filter in onFalse branch affects onTrue branch

### ISSUE THREAD

**Issue by Izhaki** [ramda/ramda#2374]
Works as expected ([REPL](http://ramdajs.com/repl/?v=0.25.0#?R.ifElse%28%0A%20%20R.isNil%2C%0A%20%20R.always%28%20%27Nil%27%20%29%2C%0A%20%20R.always%28%20%27Not%20Nil%27%20%29%0A%29%28undefined%29)):

```javascript
R.ifElse(
  R.isNil,
  R.always( 'Nil' ),
  R.always( 'Young' )
)(undefined)
//=> 'Nil'
```

Works as unexpected ([REPL](http://ramdajs.com/repl/?v=0.25.0#?R.ifElse%28%0A%20%20R.isNil%2C%0A%20%20R.always%28%20%27Nil%27%20%29%2C%0A%20%20R.filter%0A%29%28undefined%29)):
```javascript
R.ifElse(
  R.isNil,
  R.always( 'Nil' ),
  R.filter           // Only difference
)(undefined)
//=> function (t){return n.apply(this,arguments)}
```

**Comment by svozza** [comment]
I'm pretty sure this is down to the way Ramda handles the arity of the functions passed to `ifElse`. If you look [here](https://github.com/ramda/ramda/blob/v0.25.0/source/ifElse.js#L31) you'll see that the function returned will be will be curried and expecting the same amount of arguments as the function with the greatest arity of those supplied to `ifElse`. In your case, `R.filter` expects the most arguments so you get back a curried function of arity 2. I think the bug here is that the docs are misleading and say that the function returned will be unary (although I would prefer to remove the complexity around supporting functions with an artity >1).

**Comment by Izhaki** [comment]
Thanks @svozza That helps a lot.

I think this is hugely counterintuitive. There's this hidden logic that is not documented (or as you said - wrongly documented as unary).

Basically, if `condition`, `onTrue` or `onFalse` has different arity, the method won't work as expected.

I'm also not sure I understand why the returned function is curried based on the `max` rather than the arity of the `condition`?

**Comment by stefano-b2c2** [comment]
Tbh, I never use `R.ifElse` because, like you, I've been tripped up by it in the past. I find a nicely formatted ternary expression in an arrow function is clear enough.

**Comment by svozza** [comment]
Tbh, I never use R.ifElse because, like you, I've been tripped up by it in the past. I find a nicely formatted ternary expression in an arrow function is clear enough.

**Comment by CrossEye** [comment]
@Izhaki 

Sorry, I've been off for a bit, having my head examined<sup>*</sup> and I missed this.

It's not clear to me what you'd want out of your non-working example.

```hs
myFunc :: (String -> Boolean) -> 
          (String -> String) -> 
          ((String -> Boolean) -> [String] -> [String]) -> ???? 
```
If you want to return the `filter` function itself when the parameter supplied is not `nil`, then you could use `always(filter)`, although the type of this would be very strange, returning 'Nil' in one case and the `filter` function in another.  If that's not what you want, I'm not quite sure what it is you do want.

I don't really share @svozza's distaste for `ifElse`, although I am also interested in moving to unary-only.

But perhaps we should at least update the documentation to have a type something like

```hs
ifElse :: (a -> Bool) -> (a -> b) -> (a -> b) -> (a -> b)
       :: (a -> b -> Bool) -> (a -> b -> c) -> (a -> b -> c) -> (a -> b -> c)
       :: (a -> b -> c-> Bool) -> (a -> b -> c -> d) -> (a -> b -> c -> d) -> (a -> b -> c -> d)
       ...
```

----------
<sup>*</sup> So, it was just my right eye and one side of my forehead, covered in shingles bumps -- that's still part of my head!

**Comment by Izhaki** [comment]
@CrossEye 

The actual use case:

A form of 'safeFilter', where:

- if a predicate is provided, return a filter function whilst partially applying its first parameter.
- if no predicate is provided, just return `R.identity`.

(I know there are other ways to achieve this, like `R.allPass`, but the example I gave doesn't seem intuitive to me).

**Comment by Izhaki** [comment]
Whist perhaps a discussion for a different issue, the `max` strategy yields all sorts of non-intuitive results. Consider:

```javascript
R.converge( R.call, [
  R.add,
  R.add(2)
])(2)
```

In my head, this is what should happen, step by step:

```javascript
R.converge( R.call, [
  R.add(2),
  R.add(2)(2)
])
```
Then:

```javascript
R.converge( R.call, [
  R.add(2),
  4
])
```
Then:

```javascript
R.call( R.add(2), 4 )
```

```javascript
6
```
But this doesn't happen. Instead you get [[REPL](http://ramdajs.com/repl/?v=0.25.0#?R.converge%28%20R.call%2C%20%5B%0A%20%20R.add%2C%0A%20%20R.add%282%29%0A%5D%29%282%29)]:

    //=> function (t){return n.apply(this,arguments)}

**Comment by CrossEye** [comment]
> The actual use case: [...] a form of 'safeFilter'

So you want a points-free version of this?

```js
curry((pred, vals) => isNil(pred) ? vals : filter(pred, vals));
```

As you noted, there are many ways to achieve this, including

```js
ifElse(isNil, nthArg(1), filter)
```
which offers a sort of binary version of `identity` via `nthArg(1)`, or

```js
ifElse(isNil, always(identity), unary(filter))
```

which reduces `filter` to a unary function.

The difference between these two is that the former can be called either like

```js
fn(isEven, [1, 2, 3, 4, 5])
```

or like

```js
fn(isEven)([1, 2, 3, 4, 5])
```

while the latter can *only* be called the second way.

----------

All this is a result of trying to handle polyadic functions here and elsewhere.  That is much closer to normal JS behavior than is forcing everything to be unary.  But it's definitely time for Ramda to consider this anyway.

**Comment by Izhaki** [comment]
Right then. So @svozza provided the lead, and @CrossEye provided the explanation.

I think there is only one stone left unturned before this issue can be closed, and that's the documentation.

Now I may have missed a legend (have I missed a legend?), but for me this:

    *…

means 0 or more arguments of any type.

In the following signature:

    (*… → a),*… → a

There is nothing to suggest that the first and the second `*...` are of the same arity (although in the case of `call` this is rather obvious).

I personally find

```javascript
ifElse :: (a -> Bool) -> (a -> b) -> (a -> b) -> (a -> b)
       :: (a -> b -> Bool) -> (a -> b -> c) -> (a -> b -> c) -> (a -> b -> c)
       :: (a -> b -> c-> Bool) -> (a -> b -> c -> d) -> (a -> b -> c -> d) -> (a -> b -> c -> d)
       ...
```
rather verbose.

Here are some ideas:

- `n*` - `(n* → a),n* → a`
- `*n…` - `(*…n → a),*…n → a`
- `n:*…` - `(n:*… → a),n:*… → a`

Or maybe:

    1..n *… => (*… → a),*… → a

At any rate, I think that a legend is more important than the actual syntax, so perhaps just keep it `*...` but have a legend somewhere saying the arity must match?

**Comment by CrossEye** [comment]
We definitely do not have a simple legend.  I did write a [long description][ts] of how we use type signatures.

But I would like to remove every internal ellipsis from the signatures.  I think they are simply confusing.  I'm afraid I don't find any of your suggested replacements convincing, though.

This really best belongs in a separate issue though.  Please feel free to open one if you want to discuss this further.



  [ts]: https://github.com/ramda/ramda/wiki/Type-Signatures

**Comment by evan-king** [comment]
For my own signatures, I use `…` exclusively to denote variadic input, but use a number as well (i.e. `a1…aN`).  The advantage of this is that complex signatures can vary the `a` to denote flow of untyped data through stages (i.e. `a → b → c`) and the `N` can be matched elsewhere to denote expected matching of argument count.  I've never had a case with 3 different variadic matchings, so it as thus far been sufficient to switch to `M` when documenting an unrelated variadic set of arguments.

Some examples:
```js
// simple:
// args :: a1…aN → [a]
// applyThese :: [a] → (a1…aN → b) → b

// complex:
// convergeP :: (b1…bN → c) → [a1…aM → Promise(b)] → a1…aM → Promise(c)
```

**Comment by CrossEye** [comment]
@evan-king: That seems a decent syntax, but what I would prefer is what's presented in https://github.com/ramda/ramda/pull/1938#issuecomment-257752500.

**Comment by Guardiannw** [comment]
@CrossEye Is it possible to update the documentation on `ifElse` to explain that it curries until the arity of the greatest argument is reached?  This was very confusing as I passed in a function that was single arity into the `ifTrue` and one that is curried but had 2 params left to fill in the `ifFalse`.  It made things work acceptably for the `ifFalse`, but I had to adjust the `ifTrue` to accomodate.

For instance, here is an example:

```
(property, initialObject) => ifElse(equals(prop(property, initialObject)), always(dissoc(property)), set(lensProp(property)))
```

In this example, I expected this `ifElse` to return a function that when called with a value, would then return a function that, when passed in a state object, would either remove the property if it was not different than the `initialObject`'s property; or would set the value to the new value.  However, this did not work.  I had to adjust the code to allow for the arity difference between `always` and `set`.  The new code had to be like this:
```
(property, initialObject) => ifElse(equals(prop(property, initialObject)), compose(dissoc(property), nthArg(1)), set(lensProp(property)))
```
This is fine, but it was not understood from the documentation that it would curry the result until it reached the max arity of the arguments.  

So, is it possible to add that to the docs?

Thanks!
